compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp

```diff
@@ -351,6 +359,179 @@ static FailureOr<Operation *> tryRaiseToView(linalg::GenericOp linalgOp,
+//      ins(%extracted_slice_0 : tensor<...x64xf16>) outs(%2 : tensor<...x64xf16>) {
+//   ^bb0(%in: f16, %out: f16):
+//     %5 = arith.negf %in : f16
+//     linalg.yield %5 : f16 
+//   } -> tensor<...x64xf16>
+//   %inserted_slice = tensor.insert_slice %3 into %1[0, ..., 0] [..., 64] [1, ..., 1] : tensor<...x64xf16> into tensor<...x128xf16>
+//   %inserted_slice_1 = tensor.insert_slice %extracted_slice into %inserted_slice[0, ..., 64] [..., 64] [1, ..., 1] : tensor<...x64xf16> into tensor<...x128xf16>
+//
+// clang-format on
+//
+// Where the input tensor is broken down along the inner most dimension, then
+// the bottom elements are negated and the tensor is reconstructed, reversing
+// the order of the slices. This is then rewritten as
+//
+// clang-format off
+//
+//   %expanded = tensor.expand_shape %0 [[0], ..., [n-1, n]] : tensor<...x128xf16> into tensor<...x2x64xf16>
+//   %2 = linalg.generic {
+//         indexing_maps = [affine_map<(d0, ..., dn-1, dn) -> (d0, ..., dn-1, dn)>],
+//         iterator_types = [n x "parallel"]
+//      } outs(%output : tensor<...x2x64xf16>) {
+//   ^bb0(%out: f16):
+//     %i0 = linalg.index 0 : index
+//     ...
+//     %in-1 = linalg.index n-1 : index
+//     %in   = linalg.index n : index
+//     %rev = arith.subi %c1, %in-1 : index
+//     %extracted = tensor.extract %expanded[%i0, ..., %rev, %in] : tensor<...x2x64xf16>
+//     %neg = arith.negf %extracted : f16
+//     %cmp = arith.cmpi eq, %in-1, %c1 : index
+//     %sel = arith.select %cmp, %neg, %extracted : f16
+//     linalg.yield %sel : f16
+//   } -> tensor<...x2x64xf16>
+//   %collapsed = tensor.collapse_shape %2 [[0], ..., [n-1, n]] : tensor<...x2x64xf16> into tensor<...x128xf16>
+//
+// clang-format on
+static std::optional<Value>
+matchCatNegateAndSlice(tensor::InsertSliceOp insertOp) {
+  /// First match against the desired op chain.
+  auto topHalf = insertOp.getSource().getDefiningOp<tensor::ExtractSliceOp>();
+  if (!topHalf || !topHalf->hasOneUse()) {
+    return std::nullopt;
+  }
+  auto dest = insertOp.getDest().getDefiningOp<tensor::InsertSliceOp>();
+  if (!dest || !dest->hasOneUse()) {
+    return std::nullopt;
+  }
+  auto outEmpty = dest.getDest().getDefiningOp<tensor::EmptyOp>();
+  if (!outEmpty) {
+    return std::nullopt;
+  }
+
+  /// TODO: This could be extended to other unary (or in general, elementwise
+  /// operations) if the need arises.
+  auto negate = dest.getSource().getDefiningOp<linalg::GenericOp>();
+  if (!negate || !isUnaryNegate(negate)) {
+    return std::nullopt;
+  }
+
+  auto bottomHalf =
+      negate.getDpsInputs()[0].getDefiningOp<tensor::ExtractSliceOp>();
+  if (!bottomHalf || !bottomHalf->hasOneUse()) {
+    return std::nullopt;
+  }
+  Value source = topHalf.getSource();
+  if (source != bottomHalf.getSource()) {
+    return std::nullopt;
+  }
+
+  /// Require that the overall operation isn't changing the tensor shape.
+  if (source.getType() != outEmpty.getType()) {
```

Comment:
Maybe also abort on dynamic shapes, since you cant really check for equality there.

EDIT: Actually the shapes are checked explicitly below I guess. So this is OK.